File: src/leetcode/h68/TextJustification.go
package main

import (
	"strings"
)

// LeetCode 68
func fullJustify(words []string, maxWidth int) []string {
	var strArr [][]string

	line := make([]string, 0)
	lineLen := 0
	for _, word := range words {
		if lineLen == 0 {
			line = append(line, word)
			lineLen += len(word)
		} else if lineLen+len(word)+1 <= maxWidth {
			line = append(line, word)
			lineLen += len(word) + 1
		} else {
			strArr = append(strArr, line)
			line = make([]string, 0)
			line = append(line, word)
			lineLen = len(word)
		}
	}
	strArr = append(strArr, line)

	var res []string
	for i, line := range strArr {
		isLastLine := i == len(strArr)-1
		res = append(res, concatWords(line, maxWidth, isLastLine))
	}
	return res
}

func concatWords(words []string, targetWidth int, isLastLine bool) string {
	if len(words) == 1 || isLastLine {
		var builder strings.Builder
		builder.WriteString(strings.Join(words, " "))
		currLen := builder.Len()
		builder.WriteString(strings.Repeat(" ", targetWidth-currLen))
		return builder.String()
	}

	totalLen := 0
	for _, word := range words {
		totalLen += len(word)
	}
	totalGap := targetWidth - totalLen
	avgGap := totalGap / (len(words) - 1)
	remainder := totalGap % (len(words) - 1)

	var builder strings.Builder
	for i, word := range words {
		if i == len(words)-1 {
			builder.WriteString(word)
			break
		}
		builder.WriteString(word)
		spaceLen := avgGap
		if i < remainder {
			spaceLen++
		}
		whiteSpaces := strings.Repeat(" ", spaceLen)
		builder.WriteString(whiteSpaces)
	}

	return builder.String()
}

File: src/leetcode/h68/TextJustification.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.h68;

import java.util.ArrayList;
import java.util.List;

/**
 * LeetCode 68
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/09/26 08:07
 */
public class TextJustification {
    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> res = new ArrayList<>();
        List<String> candidates = new ArrayList<>();
        int lineLen = 0;
        for (String word : words) {
            if (word.length() + lineLen > maxWidth) {
                // distribute empty spaces
                StringBuilder sb = new StringBuilder();
                if (candidates.size() == 1) {
                    sb.append(candidates.get(0));
                    while (sb.length() < maxWidth) {
                        sb.append(" ");
                    }
                } else {
                    int totalSpaces = maxWidth - lineLen + candidates.size();
                    int spaces = totalSpaces / (candidates.size() - 1);
                    int remainder = totalSpaces % (candidates.size() - 1);
                    for (int i = 0; i < candidates.size() - 1; i++) {
                        sb.append(candidates.get(i));
                        for (int j = 0; j < spaces; j++) {
                            sb.append(" ");
                        }
                        if (i < remainder) {
                            sb.append(" ");
                        }
                    }
                    sb.append(candidates.get(candidates.size() - 1));
                }

                // add it to res
                res.add(sb.toString());
                // reset
                candidates.clear();
                candidates.add(word);
                lineLen = word.length() + 1;
            } else {
                candidates.add(word);
                lineLen += word.length() + 1;
            }
        }

        StringBuilder sb = new StringBuilder();
        sb.append(String.join(" ", candidates));
        while (sb.length() < maxWidth) {
            sb.append(" ");
        }
        res.add(sb.toString());

        return res;
    }
}
